README
NAME
    AnyEvent::OWNet - Client for 1-wire File System server

VERSION
    version 1.110430

SYNOPSIS
      # IMPORTANT: the API is subject to change

      my $ow = AnyEvent::OWNet->new(host => '127.0.0.1',
                                    on_error => sub { warn @_ });

      # Read temperature sensor
      $ow->read('/10.123456789012/temperature', sub { my ($res) = @_; ... });

      # List all devices
      my $cv;
      $cv = $ow->devices(sub {
                           my $dev = shift;
                           print $dev, "\n";
                         });
      $cv->recv;

      # Read the temperatures of all devices that are found
      $cv = $ow->devices(sub {
                           my $dev = shift;
                           $cv->begin;
                           $ow->get($dev.'temperature',
                                    sub {
                                      my $res = shift;
                                      $cv->end;
                                      my $value = $res->{data};
                                      return unless (defined $value);
                                      print $dev, ' = ', 0+$value, "\n";
                                    });
                         });
      $cv->recv;

      # short version of the above
      $cv = $ow->device_files(sub {
                                my ($dev, $file, $value) = @_;
                                print $dev, ' = ', 0+$value, "\n";
                              }, 'temperature');
      $cv->recv;

      # read humidity as well
      $cv = $ow->device_files(sub {
                                my ($dev, $file, $value) = @_;
                                print $dev, $file, ' = ', 0+$value, "\n";
                              }, ['temperature', 'humidity']);
      $cv->recv;

DESCRIPTION
    AnyEvent module for handling communication with an owfs 1-wire server
    daemon.

METHODS
  "new( %parameter_hash )"
    Constructs a new AnyEvent::OWNet object. The parameter hash can contain
    values for the following keys:

    "host"
        The host IP of the running "owserver" daemon. Default is the IPv4
        loopback address, 127.0.0.1.

    "port"
        The TCP port of the running "owserver" daemon. Default is 4304.

    "timeout"
        The timeout in seconds to wait for responses from the server.
        Default is 5 seconds.

  "read($path, $sub)"
    Perform an OWNet "read" operation for the given path.

  "write($path, $value, $sub)"
    Perform an OWNet "write" operation of the given value to the given path.

  "dir($path, $sub)"
    Perform an OWNet "dir" operation for the given path. The callback will
    be called once with the list of directory entries in the data field
    which isn't consistent with the (misguided?) low-latency intent of this
    operation so using dirall probably makes more sense provided the server
    supports it.

  "present($path, $sub)"
    Perform an OWNet "present" check on the given path.

  "dirall($path, $sub)"
    Perform an OWNet "dirall" operation on the given path.

  "get($path, $sub)"
    Perform an OWNet "get" operation on the given path.

  "dirallslash($path, $sub)"
    Perform an OWNet "dirall" operation on the given path.

  "getslash($path, $sub)"
    Perform an OWNet "get" operation on the given path.

  "all_cv( [ $condvar ] )"
    This method returns the AnyEvent condvar that is used to track all
    outstanding operations. It can also be used to set the initial value but
    this is only sensible when no operations are currently outstanding and
    is not normally necessary.

  "cleanup( @error )"
    This method is called on error or when the closing the connection to
    free up resources and notify any receivers of errors.

  "connect( [ $command, $callback|$condvar ] )"
    This method connects to the "owserver" daemon. It is called
    automatically when the first command is attempted.

  "devices( $callback, [ $path, [ $condvar ] ] )"
    This method identifies all devices below the given path (or '/' if the
    path is not given). An "AnyEvent" condvar may also be supplied that will
    be used to track "begin" and "end" of all actions carried out during the
    identification process. If no condvar is provided then one will be
    created. The condvar used is returned by this method.

    The supplied callback is called for each device with the path to each
    device as the first argument and the condvar for the operation as the
    second argument. The intention of passing the callback the condvar (that
    if not provided is created by the initial call) is to enable the
    callbacks that need to make further asynchronous calls to use "begin"
    calls and "end" calls (in the async callback) on the condvar so that the
    complete operation may be tracked. See the SYNOPSIS for an example.

    This method currently assumes that the "owserver" supports the
    "getslash" function and if this is not the case it will fail.

  "device_files( $callback, $file, [ $path, [ $condvar ] ] )"
    Visit each device using devices() and call the callback with the result
    of successful get() calls for $file relative to each device found. If
    $file is an array reference each array element is treated as a relative
    file.

  "anyevent_read_type()"
    This method is used to register an AnyEvent::Handle read type to read
    "OWNet" replies from an "owserver" daemon.

TODO
    The code assumes that the "owserver" supports persistence and does not
    check the response flags to recognize when it is not.

SEE ALSO
    AnyEvent(3)

    OWFS Website: http://owfs.org/

    OWFS Protocol Document: http://owfs.org/index.php?page=owserver-protocol

AUTHOR
    Mark Hindess <soft-cpan@temporalanomaly.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2011 by Mark Hindess.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension AnyEvent::OWNet.

1.110430  2011-02-12 10:27:04 Europe/London
    - Reduce perl version requirement to 5.8.

1.110180  2011-01-18 16:55:46 Europe/London
    - Add higher level file visitor method.
    - Make test less platform-specific (thx CPAN testers).

1.110050  2011-01-05 22:20:32 Europe/London
    - Initial release.


lib/AnyEvent/OWNet.pm
use strict;
use warnings;
package AnyEvent::OWNet;
BEGIN {
  $AnyEvent::OWNet::VERSION = '1.110430';
}

# ABSTRACT: Client for 1-wire File System server


use 5.008;
use constant DEBUG => $ENV{ANYEVENT_OWNET_DEBUG};
use AnyEvent;
use AnyEvent::Handle;
use AnyEvent::Socket;
use Carp qw/croak/;
use Try::Tiny;

use AnyEvent::OWNet::Constants;

use AnyEvent::OWNet::Response;


sub new {
  my ($pkg, %p) = @_;
  bless
    {
     connect_queue => [],
     host => '127.0.0.1',
     port => 4304,
     timeout => 5,
     %p,
    }, $pkg;
}

sub _msg {
  my ($self, $req) = @_;
  my $version = exists $req->{version} ? $req->{version} : 0;
  my $data = exists $req->{data} ? $req->{data} : '';
  my $payload = length $data;
  my $type =
    exists $req->{type} ? $req->{type} : OWNET_MSG_READ; # default to read
  my $sg = exists $req->{sg} ? $req->{sg} : OWNET_DEFAULT_FLAGS;
  my $size = exists $req->{size} ? $req->{size} : OWNET_DEFAULT_DATA_SIZE;
  my $offset = exists $req->{offset} ? $req->{offset} : 0;
  return pack 'N6a*', $version, $payload, $type, $sg, $size, $offset, $data;
}


sub read {
  my ($self, $path, $sub) = @_;
  $self->_run_cmd({ data => $path.chr(0), type => OWNET_MSG_READ }, $sub);
}


sub write {
  my ($self, $path, $value, $sub) = @_;
  $self->_run_cmd({ data => $path.chr(0).$value,
                   size => length $value,
                   type => OWNET_MSG_WRITE }, $sub);
}


sub dir {
  my ($self, $path, $sub) = @_;
  $self->_run_cmd({ data => $path."\0", type => OWNET_MSG_DIR, size => 0 },
                 $sub);
}


sub present {
  my ($self, $path, $sub) = @_;
  $self->_run_cmd({ data => $path."\0", type => OWNET_MSG_PRESENT }, $sub);
}


sub dirall {
  my ($self, $path, $sub) = @_;
  $self->_run_cmd({ data => $path."\0", type => OWNET_MSG_DIRALL }, $sub);
}


sub get {
  my ($self, $path, $sub) = @_;
  $self->_run_cmd({ data => $path."\0", type => OWNET_MSG_GET }, $sub);
}


sub dirallslash {
  my ($self, $path, $sub) = @_;
  $self->_run_cmd({ data => $path."\0", type => OWNET_MSG_DIRALLSLASH }, $sub);
}


sub getslash {
  my ($self, $path, $sub) = @_;
  $self->_run_cmd({ data => $path."\0", type => OWNET_MSG_GETSLASH }, $sub);
}

sub _run_cmd {
  my $self = shift;
  my $cmd  = shift;

  print STDERR 'Running command, ', $cmd->{type}, "\n" if DEBUG;
  $self->{cmd_cb} or return $self->connect($cmd, @_);
  $self->{cmd_cb}->($cmd, @_);
}

sub DESTROY { }


sub all_cv {
  my $self = shift;
  $self->{all_cv} = shift if @_;
  unless ($self->{all_cv}) {
    $self->{all_cv} = AnyEvent->condvar;
  }
  $self->{all_cv};
}


sub cleanup {
  my $self = shift;
  print STDERR "cleanup\n" if DEBUG;
  $self->{all_cv}->croak(@_) if ($self->{all_cv});
  while (@{$self->{connect_queue}}) {
    my $queue = shift @{$self->{connect_queue}};
    my($cv, @args) = @$queue;
    $cv->croak(@_);
  }
  delete $self->{all_cv};
  delete $self->{cmd_cb};
  delete $self->{sock};
  $self->{on_error}->(@_) if $self->{on_error};
}


sub connect {
  my $self = shift;

  my $cv;
  if (@_) {
    $cv = AnyEvent->condvar;
    push @{$self->{connect_queue}}, [ $cv, @_ ];
  }

  return $cv if $self->{sock};

  $self->{sock} = tcp_connect $self->{host}, $self->{port}, sub {

    my $fh = shift
      or do {
        my $err = "Can't connect owserver: $!";
        $self->cleanup($err);
        $cv->croak($err);
        return
      };

    warn "Connected\n" if DEBUG;

    my $hd =
      AnyEvent::Handle->new(
                            fh => $fh,
                            on_error => sub {
                              print STDERR "handle error $_[2]\n" if DEBUG;
                              $_[0]->destroy;
                              if ($_[1]) {
                                $self->cleanup('Error: '.$_[2]);
                              }
                            },
                            on_eof => sub {
                              print STDERR "handle eof\n" if DEBUG;
                              $_[0]->destroy;
                              $self->cleanup('Connection closed');
                            },
                            on_timeout => sub {
                              print STDERR "handle timeout\n" if DEBUG;
                              $_[0]->destroy;
                              $self->cleanup('Socket timeout');
                            }
                           );
    $self->{cmd_cb} = sub {
      $self->all_cv->begin;
      my $command = shift;

      my ($cv, $cb);
      if (@_) {
        $cv = pop if UNIVERSAL::isa($_[-1], 'AnyEvent::CondVar');
        $cb = pop if ref $_[-1] eq 'CODE';
      }

      my $msg = $self->_msg($command);
      print STDERR "sending command ", $command->{type}, "\n" if DEBUG;
      warn 'Sending: ', (unpack 'H*', $msg), "\n" if DEBUG;

      $hd->push_write($msg);
      $hd->timeout($self->{timeout});

      $cv ||= AnyEvent->condvar;

      print STDERR "using condvar $cv\n" if DEBUG;

      $cv->cb(sub {
                my $cv = shift;
                print STDERR "calling callback $cv\n" if DEBUG;
                try {
                  my $res = $cv->recv;
                  $cb->($res);
                } catch {
                  ($self->{on_error} || sub { die "ARGH: @_\n"; })->($_);
                }
              }) if $cb;

      $hd->push_read(ref $self, $command => sub {
                       my($handle, $res, $err) = @_;
                       $hd->timeout(0);
                       print STDERR "read finished $cv\n" if DEBUG;
                       print STDERR "read ",
                         ($cv->ready ? "ready" : "not ready"), "\n" if DEBUG;
                       $self->all_cv->end;
                       if ($err) {
                         print STDERR "returning error $err\n" if DEBUG;
                         return $cv->croak($res)
                       }
                       print STDERR "Sending $res\n" if DEBUG;
                       $cv->send($res);
                     });
      return $cv;
    };

    while (@{$self->{connect_queue}}) {
      my $queue = shift @{$self->{connect_queue}};
      my($cv, @args) = @$queue;
      $self->{cmd_cb}->(@args, $cv);
    }
#    $cv->send(1);
  };

  return $cv;
}


sub devices {
  my ($self, $cb, $offset, $cv) = @_;
  $offset ||= '/';
  $cv ||= AnyEvent->condvar;
  print STDERR "devices: $offset\n" if DEBUG;
  $cv->begin;
  $self->getslash($offset, sub {
                    my $res = shift;
                    if ($res->is_success) {
                      foreach my $d ($res->data_list) {
                        if ($d =~ m!^.*/[0-9a-f]{2}\.[0-9a-f]{12}/$!i) {
                          $cb->($d, $cv);
                          $self->devices($cb, $d, $cv);
                        } elsif ($d =~ m!/(?:main|aux)/$!) {
                          $self->devices($cb, $d, $cv);
                        }
                      }
                    } # TOFIX: propogate error?
                    $cv->end;
                  });
  $cv;
}


sub device_files {
  my ($self, $cb, $files, $offset, $cv) = @_;
  $files = [$files] unless (ref $files);
  $cv = $self->devices(sub {
                 my $dev = shift;
                 foreach my $file (@$files) {
                   $cv->begin;
                   $self->get($dev.$file,
                            sub {
                              my $res = shift;
                              $cv->end;
                              my $value = $res->{data};
                              return unless (defined $value);
                              $cb->($dev, $file, 0+$value);
                            });
                 }
               }, $offset, $cv);
}


sub anyevent_read_type {
  my ($handle, $cb, $command) = @_;

  my $MAX_RETURN = 66000;
  my @data;
  sub {
    my $rbuf = \$handle->{rbuf};

  REDO:
    return unless (defined $$rbuf);
    my $len;

    my %result;
    my $header;
    do {
      $len = length $$rbuf;
      print STDERR "read_type has $len bytes\n" if DEBUG;
      print STDERR "read_type has ", (unpack 'H*', $$rbuf), "\n" if DEBUG;
      return unless ($len >= 24);
      @result{qw/version payload ret sg size offset/} = unpack 'N6', $$rbuf;
      $header = substr $$rbuf, 0, 24, '';
      print STDERR "read_type header ", (unpack 'H*', $header), "\n" if DEBUG;
      if ($result{'ret'} > $MAX_RETURN) {
        $cb->($handle, AnyEvent::OWNet::Response->new(%result));
        return 1;
      }
    } while ($result{payload} > $MAX_RETURN);

    my $total_len = 24 + $result{payload};
    print STDERR "read_type have ", $len, " need ", $total_len, "\n" if DEBUG;
    unless ($len >= $total_len) {
      $$rbuf = $header.$$rbuf;
      return;
    }

    my $data = substr $$rbuf, 0, $result{payload}, '';
    if ($command->{type} == OWNET_MSG_DIR) {
      if ($data eq '') {
        $result{data} = \@data;
      } else {
        push @data, substr $data, 0, -1;
        goto REDO;
      }
    } else {
      $result{data} = $data;
    }
    print STDERR "read_type complete\n" if DEBUG;
    $cb->($handle, AnyEvent::OWNet::Response->new(%result));
    return 1;
  }
}

1;


__END__
=pod

=head1 NAME

AnyEvent::OWNet - Client for 1-wire File System server

=head1 VERSION

version 1.110430

=head1 SYNOPSIS

  # IMPORTANT: the API is subject to change

  my $ow = AnyEvent::OWNet->new(host => '127.0.0.1',
                                on_error => sub { warn @_ });

  # Read temperature sensor
  $ow->read('/10.123456789012/temperature', sub { my ($res) = @_; ... });

  # List all devices
  my $cv;
  $cv = $ow->devices(sub {
                       my $dev = shift;
                       print $dev, "\n";
                     });
  $cv->recv;

  # Read the temperatures of all devices that are found
  $cv = $ow->devices(sub {
                       my $dev = shift;
                       $cv->begin;
                       $ow->get($dev.'temperature',
                                sub {
                                  my $res = shift;
                                  $cv->end;
                                  my $value = $res->{data};
                                  return unless (defined $value);
                                  print $dev, ' = ', 0+$value, "\n";
                                });
                     });
  $cv->recv;

  # short version of the above
  $cv = $ow->device_files(sub {
                            my ($dev, $file, $value) = @_;
                            print $dev, ' = ', 0+$value, "\n";
                          }, 'temperature');
  $cv->recv;

  # read humidity as well
  $cv = $ow->device_files(sub {
                            my ($dev, $file, $value) = @_;
                            print $dev, $file, ' = ', 0+$value, "\n";
                          }, ['temperature', 'humidity']);
  $cv->recv;

=head1 DESCRIPTION

AnyEvent module for handling communication with an owfs 1-wire server
daemon.

=head1 METHODS

=head2 C<new( %parameter_hash )>

Constructs a new L<AnyEvent::OWNet> object.  The parameter hash can contain
values for the following keys:

=over

=item C<host>

The host IP of the running C<owserver> daemon.  Default is the IPv4
loopback address, C<127.0.0.1>.

=item C<port>

The TCP port of the running C<owserver> daemon.  Default is C<4304>.

=item C<timeout>

The timeout in seconds to wait for responses from the server.  Default
is 5 seconds.

=back

=head2 C<read($path, $sub)>

Perform an OWNet C<read> operation for the given path.

=head2 C<write($path, $value, $sub)>

Perform an OWNet C<write> operation of the given value to the given path.

=head2 C<dir($path, $sub)>

Perform an OWNet C<dir> operation for the given path.  The callback
will be called once with the list of directory entries in the data
field which isn't consistent with the (misguided?) low-latency intent
of this operation so using L<dirall> probably makes more sense
provided the server supports it.

=head2 C<present($path, $sub)>

Perform an OWNet C<present> check on the given path.

=head2 C<dirall($path, $sub)>

Perform an OWNet C<dirall> operation on the given path.

=head2 C<get($path, $sub)>

Perform an OWNet C<get> operation on the given path.

=head2 C<dirallslash($path, $sub)>

Perform an OWNet C<dirall> operation on the given path.

=head2 C<getslash($path, $sub)>

Perform an OWNet C<get> operation on the given path.

=head2 C<all_cv( [ $condvar ] )>

This method returns the L<AnyEvent> condvar that is used to track all
outstanding operations.  It can also be used to set the initial value
but this is only sensible when no operations are currently outstanding
and is not normally necessary.

=head2 C<cleanup( @error )>

This method is called on error or when the closing the connection to
free up resources and notify any receivers of errors.

=head2 C<connect( [ $command, $callback|$condvar ] )>

This method connects to the C<owserver> daemon.  It is called automatically
when the first command is attempted.

=head2 C<devices( $callback, [ $path, [ $condvar ] ] )>

This method identifies all devices below the given path (or '/' if the
path is not given).  An C<AnyEvent> condvar may also be supplied that
will be used to track C<begin> and C<end> of all actions carried out
during the identification process.  If no condvar is provided then one
will be created.  The condvar used is returned by this method.

The supplied callback is called for each device with the path to each
device as the first argument and the condvar for the operation as the
second argument.  The intention of passing the callback the condvar
(that if not provided is created by the initial call) is to enable the
callbacks that need to make further asynchronous calls to use C<begin>
calls and C<end> calls (in the async callback) on the condvar so that
the complete operation may be tracked.  See the L<SYNOPSIS> for an
example.

This method currently assumes that the C<owserver> supports the C<getslash>
function and if this is not the case it will fail.

=head2 C<device_files( $callback, $file, [ $path, [ $condvar ] ] )>

Visit each device using L<devices()> and call the callback with the
result of successful L<get()> calls for C<$file> relative to each
device found.  If C<$file> is an array reference each array element
is treated as a relative file.

=head2 C<anyevent_read_type()>

This method is used to register an L<AnyEvent::Handle> read type
to read C<OWNet> replies from an C<owserver> daemon.

=head1 TODO

The code assumes that the C<owserver> supports persistence and does
not check the response flags to recognize when it is not.

=head1 SEE ALSO

AnyEvent(3)

OWFS Website: http://owfs.org/

OWFS Protocol Document: http://owfs.org/index.php?page=owserver-protocol

=head1 AUTHOR

Mark Hindess <soft-cpan@temporalanomaly.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Mark Hindess.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/AnyEvent/OWNet/Response.pm
use strict;
use warnings;
package AnyEvent::OWNet::Response;
BEGIN {
  $AnyEvent::OWNet::Response::VERSION = '1.110430';
}

# ABSTRACT: Module for responses from 1-wire File System server


sub new {
  my ($pkg, %p) = @_;
  bless { %p }, $pkg;
}


sub is_success {
  shift->{ret} == 0
}


sub return_code {
  shift->{ret}
}


sub version {
  shift->{version}
}


sub flags {
  shift->{sg}
}


sub payload_length {
  shift->{payload}
}


sub size {
  shift->{size}
}


sub offset {
  shift->{offset}
}


sub data_list {
  my $self = shift;
  unless (ref $self->{data}) {
    $self->{data} = [ split /,/, substr $self->{data}, 0, -1 ];
  }
  @{$self->{data}}
}


sub data {
  shift->{data}
}

1;

__END__
=pod

=head1 NAME

AnyEvent::OWNet::Response - Module for responses from 1-wire File System server

=head1 VERSION

version 1.110430

=head1 SYNOPSIS

  # normally instantiated by AnyEvent::OWNet command methods

=head1 DESCRIPTION

Module to represent responses from owfs 1-wire server daemon.

=head1 METHODS

=head2 C<new()>

Constructs a new L<AnyEvent::OWNet::Response> object.  It is called by
L<AnyEvent::OWNet> in response to messages received from the
C<owserver> daemon.

=head2 C<is_success()>

Returns true if the response object represents a successful request.

=head2 C<return_code()>

Returns the return code of the response from the C<owserver> daemon.

=head2 C<version()>

Returns the protocol version number of the response from the
C<owserver> daemon.

=head2 C<flags()>

Returns the flags field of the response from the C<owserver> daemon.
The L<AnyEvent::OWNet::Constants#ownet_temperature_units>,
L<AnyEvent::OWNet::Constants#ownet_pressure_units>, and
L<AnyEvent::OWNet::Constants#ownet_display_format> functions can be
used to extract some elements from this value.

=head2 C<payload_length()>

Returns the payload length field of the response from the C<owserver>
daemon.

=head2 C<size()>

Returns the size of the data element of the response from the
C<owserver> daemon.

=head2 C<offset()>

Returns the offset field of the response from the C<owserver> daemon.

=head2 C<data_list()>

Returns the data from the response as a list.  This is a intend for use
when the response corresponds to a directory listing request.

=head2 C<data()>

Returns the data from the response as a scalar.  This is a intend for
use when the response corresponds to a file C<read>.  However, it
returns the raw data for any request so while for a C<read> it may be
a simple scalar value, it may also be a comma separated list (e.g. for
a C<dirall> request) or an array reference (e.g. for a C<dir> request).

=head1 AUTHOR

Mark Hindess <soft-cpan@temporalanomaly.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Mark Hindess.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/AnyEvent/OWNet/Constants.pm
use strict;
use warnings;
package AnyEvent::OWNet::Constants;
BEGIN {
  $AnyEvent::OWNet::Constants::VERSION = '1.110430';
}

# ABSTRACT: Module to export constants for 1-wire File System daemon protocol


my %constants =
  (
   OWNET_BUS_RET     => 0x00000002,
   OWNET_PERSISTENT  => 0x00000004,
   OWNET_ALIAS       => 0x00000008,
   OWNET_SAFEMODE    => 0x00000010,
   OWNET_NET         => 0x00000100,

   OWNET_CENTIGRADE  => 0x00000000,
   OWNET_FAHRENHEIT  => 0x00010000,
   OWNET_KELVIN      => 0x00020000,
   OWNET_RANKINE     => 0x00030000,
   OWNET_TEMP_MASK   => 0x00030000,

   OWNET_MILLIBAR      => 0x00000000,
   OWNET_ATMOSPHERE    => 0x00040000,
   OWNET_MM_MERCURY    => 0x00080000,
   OWNET_IN_MERCURY    => 0x000C0000,
   OWNET_PSI           => 0x00100000,
   OWNET_PASCAL        => 0x00140000,
   OWNET_PRESSURE_MASK => 0x001C0000,

   OWNET_DISP_F_I    => 0x00000000, # f.i    e.g. /10.67C6697351FF
   OWNET_DISP_FI     => 0x01000000, # fi     e.g. /1067C6697351FF
   OWNET_DISP_F_I_C  => 0x02000000, # f.i.c  e.g. /10.67C6697351FF.8D
   OWNET_DISP_F_IC   => 0x03000000, # f.ic   e.g. /10.67C6697351FF8D
   OWNET_DISP_FI_C   => 0x04000000, # fi.c   e.g. /10.67C6697351FF8D
   OWNET_DISP_FIC    => 0x05000000, # fic    e.g. /1067C6697351FF8D
   OWNET_DISP_MASK   => 0x07000000,

   OWNET_MSG_NOP         => 0x1, # deprecated
   OWNET_MSG_READ        => 0x2,
   OWNET_MSG_WRITE       => 0x3,
   OWNET_MSG_DIR         => 0x4,
   OWNET_MSG_SIZE        => 0x5, # deprecated
   OWNET_MSG_PRESENT     => 0x6,
   OWNET_MSG_DIRALL      => 0x7,
   OWNET_MSG_GET         => 0x8,
   OWNET_MSG_DIRALLSLASH => 0x9,
   OWNET_MSG_GETSLASH    => 0xa,

   OWNET_DEFAULT_DATA_SIZE => 0x80e8,
  );

$constants{OWNET_DEFAULT_FLAGS} =
  $constants{OWNET_NET} | $constants{OWNET_BUS_RET} |
  $constants{OWNET_ALIAS} | $constants{OWNET_PERSISTENT};

sub import {
  no strict qw/refs/; ## no critic
  my $pkg = caller(0);
  foreach (keys %constants) {
    my $v = $constants{$_};
    *{$pkg.'::'.$_} = sub () { $v };
  }
  *{$pkg.'::ownet_temperature_units'} = \&ownet_temperature_units;
  *{$pkg.'::ownet_pressure_units'} = \&ownet_pressure_units;
  *{$pkg.'::ownet_display_format'} = \&ownet_display_format;
}


sub ownet_temperature_units {
  my $flag = shift;
  [qw/C F K R/]->[($flag & $constants{OWNET_TEMP_MASK}) >> 16]
}


sub ownet_pressure_units {
  my $flag = shift;
  [qw/mbar atm mmHg
      inHg psi Pa/]->[($flag & $constants{OWNET_PRESSURE_MASK}) >> 18]
}


sub ownet_display_format {
  my $flag = shift;
  [qw/f.i fi f.i.c
      f.ic fi.c fic/]->[($flag & $constants{OWNET_DISP_MASK}) >> 24]
}

1;

__END__
=pod

=head1 NAME

AnyEvent::OWNet::Constants - Module to export constants for 1-wire File System daemon protocol

=head1 VERSION

version 1.110430

=head1 SYNOPSIS

  use AnyEvent::OWNet::Constants;

=head1 DESCRIPTION

Module to export constants for owfs daemon protocol.

=head1 C<FUNCTIONS>

=head2 C<ownet_temperature_units( $flags )>

Returns the temperature units for the given flags from an
L<AnyEvent::OWNet::Response> flags attribute.

=head2 C<ownet_pressure_units( $flags )>

Returns the pressure units for the given flags from an
L<AnyEvent::OWNet::Response> flags attribute.

=head2 C<ownet_display_format( $flags )>

Returns the display format from the given flags of an
L<AnyEvent::OWNet::Response> flags attribute.

=head1 AUTHOR

Mark Hindess <soft-cpan@temporalanomaly.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Mark Hindess.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Makefile.PL

use strict;
use warnings;

BEGIN { require 5.008; }

use ExtUtils::MakeMaker 6.31;



my %WriteMakefileArgs = (
  'ABSTRACT' => 'Client for 1-wire File System server',
  'AUTHOR' => 'Mark Hindess <soft-cpan@temporalanomaly.com>',
  'BUILD_REQUIRES' => {
    'English' => '0',
    'Exporter' => '0',
    'File::Temp' => '0',
    'Test::More' => '0'
  },
  'CONFIGURE_REQUIRES' => {
    'ExtUtils::MakeMaker' => '6.31'
  },
  'DISTNAME' => 'AnyEvent-OWNet',
  'EXE_FILES' => [],
  'LICENSE' => 'perl',
  'NAME' => 'AnyEvent::OWNet',
  'PREREQ_PM' => {
    'AnyEvent' => '0',
    'AnyEvent::Handle' => '0',
    'AnyEvent::Socket' => '0',
    'Carp' => '0',
    'Try::Tiny' => '0',
    'constant' => '0'
  },
  'VERSION' => '1.110430',
  'test' => {
    'TESTS' => 't/*.t'
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



